src/main/java/de/tum/cit/aet/artemis/core/service/RateLimitService.java

```diff
@@ -0,0 +1,104 @@
+package de.tum.cit.aet.artemis.core.service;
+
+import static de.tum.cit.aet.artemis.core.config.Constants.PROFILE_CORE;
+import static de.tum.cit.aet.artemis.core.util.HttpRequestUtils.getIpStringFromRequest;
+
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+import jakarta.servlet.http.HttpServletRequest;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.context.annotation.Lazy;
+import org.springframework.context.annotation.Profile;
+import org.springframework.stereotype.Service;
+import org.springframework.web.context.request.RequestAttributes;
+import org.springframework.web.context.request.RequestContextHolder;
+
+import de.tum.cit.aet.artemis.core.config.RateLimitConfig;
+import de.tum.cit.aet.artemis.core.exception.RateLimitExceededException;
+import de.tum.cit.aet.artemis.core.security.RateLimitType;
+import inet.ipaddr.IPAddress;
+import inet.ipaddr.IPAddressString;
+import io.github.bucket4j.Bucket;
+import io.github.bucket4j.BucketConfiguration;
+import io.github.bucket4j.ConsumptionProbe;
+import io.github.bucket4j.grid.hazelcast.HazelcastProxyManager;
+
+@Profile(PROFILE_CORE)
+@Service
+@Lazy
+public class RateLimitService {
+
+    private static final Logger log = LoggerFactory.getLogger(RateLimitService.class);
+
+    private final HazelcastProxyManager<String> proxyManager;
+
+    private final Map<Integer, BucketConfiguration> perMinuteCfgCache = new ConcurrentHashMap<>();
```

Comment:
Would it make sense to use a Hazelcast Map here to sync this automatically between all core nodes?